### PR TITLE
Restores linters and separate kinds of tests

### PR DESCRIPTION
**Pull request by JekaMas**
1. Restores and updates linters
2. Introduce a command to check reproducible builds (should be fixed in a separate issue)
3. Introduce --shuffle for unit tests
4. Introduce --race for unit tests
5. Separate e2e tests
6. Adds build info into the binary for future healthcheck
7. Adds an example of goleak test - `TestSimulatedBackend`
8. Adds `make test-race`, however switched off in CI. Fix data races should be done package by package as a bunch of new issues.

The plan is to not fix all linters at the moment but rather review the PR, force-merge it and then fix linters only for
1. New PRs
2. Packages owned by Polygon team, eg `consensus/bor`

### CODE REVIEW

**Comment by cffls** [comment]
Nice work! Looks pretty good overall. The linter failed in CI, but my browser couldn't load the page, probably because there were too many lines printed by linter.

File: .github/workflows/ci.yml
Comment (JekaMas) [reply .github/workflows/ci.yml:6]: This stuff and above is to
1. Run CI on PRs and new commits
2. Run CI on master branch after any merge into master.

File: Makefile
Comment (JekaMas) [reply Makefile:60]: What's the purpose?

File: .github/workflows/ci.yml
Comment (cffls) [reply .github/workflows/ci.yml:6]: Got it. Would this achieve the same effect? When a PR is merged to master, it will trigger a push event (see [this](https://github.community/t/trigger-workflow-only-on-pull-request-merge/17359/3)).

```
  pull_request:
    branches:
      - master # It might be even better to just use "*" , because we usually do PR against a feature branch
    types: [opened, synchronize]

  push:
    branches:
      - "**"
```

File: Makefile
Comment (cffls) [reply Makefile:60]: Since we will fail CI when linter fails anyway, I think it may be more convenient to combine unit tests and linter into one command. I am also fine with separating them. It is your call. :D

File: .github/workflows/ci.yml
Comment (JekaMas) [reply .github/workflows/ci.yml]: To run test on master only if a PR is closed and merged.

File: .github/workflows/ci.yml
Comment (JekaMas) [reply .github/workflows/ci.yml:6]: Both solutions were a bit incorrect causing 2 CI processes at the same time. I think I fixed it.

File: Makefile
Comment (JekaMas) [reply Makefile:60]: It should we 'green' now.

File: .github/workflows/ci.yml
Comment (cffls) [reply .github/workflows/ci.yml:6]: This works when a PR is created from a branch in maticnetwork/bor, but it doesn't seem to run any CI test if the PR is created from a fork repo of bor. I added a new dummy commit on top of your branch, pushed this branch to my fork of bor, and created [this PR](https://github.com/maticnetwork/bor/pull/400). The CI in this PR was skipped, which is not a desired behavior.

Comment (JekaMas) [reply .github/workflows/ci.yml]: That solution triggers 2 CI processes for each commit.

**Review by 0xsharma — APPROVED** [review]
LGTM 🎉

**Review by cffls — APPROVED** [review]
LGTM, thanks!

**Review by temaniarpit27 — APPROVED** [review]

**Pull request merged**
